lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,184 @@
+//===- DropConst.cpp - Check and remove const types -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DropConst pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+
+#define DEBUG_TYPE "drop-const"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+
+class OpVisitor : public FIRRTLVisitor<OpVisitor> {
+public:
+  using FIRRTLVisitor<OpVisitor>::visitExpr;
+  using FIRRTLVisitor<OpVisitor>::visitDecl;
+  using FIRRTLVisitor<OpVisitor>::visitStmt;
+
+  LogicalResult checkModule(FModuleLike module) {
+    auto fmodule = dyn_cast<FModuleOp>(*module);
+    // Find 'const' ports
+    auto portTypes = SmallVector<Attribute>(module.getPortTypes());
+    for (size_t portIndex = 0, numPorts = module.getPortTypes().size();
+         portIndex < numPorts; ++portIndex) {
+      if (auto convertedType = convertType(module.getPortType(portIndex))) {
+        // If this is an FModuleOp, register the block argument to drop 'const'
+        if (fmodule)
+          constValuesToConvert.push_back(
+              {fmodule.getArgument(portIndex), convertedType});
+        portTypes[portIndex] = TypeAttr::get(convertedType);
+      }
+    }
+
+    // Update the module signature with non-'const' ports
+    module->setAttr(FModuleLike::getPortTypesAttrName(),
+                    ArrayAttr::get(module.getContext(), portTypes));
+
+    if (!fmodule)
+      return result;
+
+    // Check the module body
+    visitDecl(fmodule);
+
+    // Drop 'const' from all registered values
+    for (auto [value, type] : constValuesToConvert)
+      value.setType(type);
+
+    return result;
+  }
+
+  void visitDecl(FModuleOp module) {
+    for (auto &op : llvm::make_early_inc_range(*module.getBodyBlock()))
+      dispatchVisitor(&op);
+  }
```

Comment:
I do not love post-processing to remove the constness.  I think your'e doing this so that you do not alter the WhenOps' condition types?  If you move the checking to a verifier, you should be able to fix this.  Alternatively, you should be able to do this in a single pass if you walk the IR in reverse.  Make sure you check out the walk API as well: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Operation.h#L724

Reply:
I'll look into this

Reply:
Updated to set the types to non-const in a single pass